lib/web_ui/lib/src/engine/text/measurement.dart

```diff
@@ -123,5 +127,248 @@ double measureSubstring(
+
+    final double textAscent = excludingTrailingSpaces.actualBoundingBoxAscent!.toDouble();
+    final double textDescent = excludingTrailingSpaces.actualBoundingBoxDescent!.toDouble();
+
+    // Divide the extra height equally between ascent and descent.
+    final double extraHeight = height - (textAscent + textDescent);
+    final double ascent = textAscent + extraHeight / 2;
+    final double descent = height - ascent;
+
+    return TextMetrics._(
+      ascent: ascent,
+      descent: descent,
+      textAscent: textAscent,
+      textDescent: textDescent,
+      widthExcludingTrailingSpaces:
+          _roundWidth(excludingTrailingSpaces.width!.toDouble()),
+      widthIncludingTrailingSpaces:
+          _roundWidth(includingTrailingSpaces.width!.toDouble()),
+    );
+  }
+
+  factory TextMetrics.forPlaceholder({
+    required PlaceholderSpan placeholder,
+    required double ascent,
+    required double descent,
+  }) {
+    assert(placeholder.height == ascent + descent);
+    return TextMetrics._(
+      ascent: ascent,
+      descent: descent,
+      textAscent: ascent,
+      textDescent: descent,
+      widthExcludingTrailingSpaces: placeholder.width,
+      widthIncludingTrailingSpaces: placeholder.width,
+    );
+  }
+
+  static const TextMetrics zero = TextMetrics._(
+    ascent: 0,
+    descent: 0,
+    textAscent: 0,
+    textDescent: 0,
+    widthExcludingTrailingSpaces: 0,
+    widthIncludingTrailingSpaces: 0,
+  );
+
+  /// The rise from the baseline as calculated from the font and style for this text.
+  final double ascent;
+
+  /// The drop from the baseline as calculated from the font and style for this text.
+  final double descent;
+
+  /// The rise from the baseline to the top of the bounding box of the measured text.
+  final double textAscent;
+
+  /// The drop from the baseline to the bottom of the bounding box of the measured text.
+  final double textDescent;
+
+  /// The width of the measured text, not including trailing spaces.
+  final double widthExcludingTrailingSpaces;
+
+  /// The width of the measured text, including any trailing spaces.
+  final double widthIncludingTrailingSpaces;
+
+  /// The width of trailing spaces in the measured text.
+  double get widthOfTrailingSpaces => widthIncludingTrailingSpaces - widthExcludingTrailingSpaces;
+
+  /// The total height as calculated from the font and style for this text.
+  double get height => ascent + descent;
+
+  /// The height of the measured text's bounding box.
+  double get textHeight => textAscent + textDescent;
+}
+
+/// Helper class that measures text in isolation and provides caching for better
+/// performance.
+class CachedMeasurements {
+  CachedMeasurements(this._canvasContext);
+
+  final DomCanvasRenderingContext2D _canvasContext;
+
+  /// A cache that maps: style -> text -> TextMetrics
+  final Map<String, Map<String, TextMetrics>> _cache =
+      <String, Map<String, TextMetrics>>{};
```

Comment:
This could keep growing indefinitely. Maybe use an LRU cache? I remember implementing a heuristic that worked well for the DOM ruler. It evicted old and unused text, but also avoided accumulating one-off measurements, e.g. a single piece of text that animated its style.